File: app/src/main/java/io/github/ovso/psytest/ui/main/MainPresenterImpl.java
package io.github.ovso.psytest.ui.main;

import io.github.ovso.psytest.R;
import io.github.ovso.psytest.ui.main.adapter.MainAdapterDataModel;
import io.github.ovso.psytest.utils.ResourceProvider;

public class MainPresenterImpl implements MainPresenter {

  private MainPresenter.View view;
  private ResourceProvider resourceProvider;
  private MainAdapterDataModel adapterDataModel;

  public MainPresenterImpl(MainPresenter.View $view, ResourceProvider $ResourceProvider,
      MainAdapterDataModel $adapterDataModel) {
    view = $view;
    resourceProvider = $ResourceProvider;
    adapterDataModel = $adapterDataModel;
    $view.changeTheme();
  }

  @Override public void onCreate() {
    view.showTitle(resourceProvider.getString(R.string.app_name));
    view.setupView();
    view.setupTabLayout();
    view.setupViewPager();
    adapterDataModel.createItems(resourceProvider.getStringArray(R.array.tabs).length);
    view.refresh();
    showTabNames();
    view.showBannerAd();
  }

  private void showTabNames() {
    String[] tabNames = resourceProvider.getStringArray(R.array.tabs);
    for (int i = 0; i < tabNames.length; i++) {
      view.showTabName(i, tabNames[i]);
    }
  }
}

File: app/src/main/java/io/github/ovso/psytest/ui/main/MainPresenterImpl.kt
package io.github.ovso.psytest.ui.main

import android.arch.lifecycle.Lifecycle.Event.ON_DESTROY
import android.arch.lifecycle.OnLifecycleEvent
import io.github.ovso.psytest.R
import io.github.ovso.psytest.R.array
import io.github.ovso.psytest.exts.plusAssign
import io.github.ovso.psytest.ui.main.adapter.MainItem
import io.github.ovso.psytest.utils.ResourceProvider
import io.github.ovso.psytest.utils.SchedulersFacade
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.BiFunction

class MainPresenterImpl(
  private val view: MainPresenter.View,
  private val resourceProvider: ResourceProvider,
  private val schedulers: SchedulersFacade
) : MainPresenter {

  private val compositeDisposable = CompositeDisposable()

  override fun onCreate() {
    view.showTitle(resourceProvider.getString(R.string.app_name))
    view.setupRv()
    view.loadAd()
    fetchList()
  }

  private fun fetchList() {
    val titles = Observable.fromIterable(resourceProvider.getStringArray(array.tabs).toList())
    val queries = Observable.fromIterable(resourceProvider.getStringArray(array.queries).toList())

    fun onSuccess(items: List<MainItem>) {
      view.submitList(items)
    }

    fun onFailure(t: Throwable) {
      println(t)
    }

    compositeDisposable += titles
        .zipWith(
            queries, BiFunction<String, String, MainItem> { name, query -> MainItem(name, query) }
        )
        .subscribeOn(schedulers.io())
        .observeOn(schedulers.ui())
        .toList()
        .subscribe(::onSuccess, ::onFailure)
  }

  @OnLifecycleEvent(ON_DESTROY)
  private fun onDestroy() {
    compositeDisposable.clear()
  }
}